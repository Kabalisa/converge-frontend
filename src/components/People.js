import React from 'react';
import PropTypes from 'prop-types';
import EditUser from './EditUser';
import DeleteButton from '../assets/images/delete.svg';
import ProfileImg from '../assets/images/profile_icon.svg';

const People = ({
  people: {
    name, accessLevel, location,
  },
}) => (
  <tr>
    <td><img className="profilePic" src={ProfileImg} alt="profilePicture" />{name}</td>
    <td>{location}</td>
    <td>{accessLevel}</td>
    <td>
      <EditUser
        userName={name}
        userLocation={location}
        accessLevel={accessLevel}
      />
      <button><img src={DeleteButton} alt="Delete" /></button>
    </td>
  </tr>
);
People.propTypes = {
  people: PropTypes.shape({
    name: PropTypes.string.isRequired,
    location: PropTypes.string.isRequired,
    accessLevel: PropTypes.string.isRequired,
  }).isRequired,
};
export default People;
