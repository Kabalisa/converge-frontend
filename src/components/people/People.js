import React from 'react';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import { IconMenu, MenuItem } from 'react-toolbox/lib/menu';
import notification from '../../utils/notification';

/**
 * Shows the menuCart besides the Access Level on the people's table
 *
 * @returns {JSX}
 */
const accessMenuCaret = () => <span className="access-by-caret" />;

/**
 * Component that shows people and their attributes
 *
 * @param {Object} people - name, email, accessLevel, location, and picture
 * @param {Object} allRoles
 * @param {function} editRole
 *
 * @returns {void}
 */
const People = ({
  people: {
    email, name, accessLevel, location, picture,
  },
  allRoles,
  refetch,
  currentPage,
  editRole,
}) => {
  const editRoleFunction = (roleId) => {
    const variables = { variables: { email, roleId } };
    editRole(variables)
      .then(() => {
        notification(
          toastr,
          'success',
          `'${name}' role has been changed successfully`,
        )();
        refetch({ page: currentPage });
      })
      .catch((err) => {
        notification(toastr, 'error', err.graphQLErrors[0].message)();
      });
  };
  return (
    <div className="table__row">
      <span>
        <img className="profilePic" src={picture} alt="profilePicture" />
        {name}
      </span>
      <span>{location}</span>
      <span>
        <span>
          {accessLevel}
          <IconMenu
            position="topRight"
            className="people-access-dropdown"
            icon={accessMenuCaret()}
          >
            {allRoles.map(role => (
              <MenuItem
                className={`access-menu ${
                  role.role === accessLevel ? 'selected' : ''
                }`}
                key={role.id}
                caption={role.role}
                onClick={() => editRoleFunction(role.id)}
              />
            ))}
          </IconMenu>
        </span>
      </span>
    </div>
  );
};

People.propTypes = {
  people: PropTypes.shape({
    email: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    location: PropTypes.string,
    accessLevel: PropTypes.string.isRequired,
    picture: PropTypes.string.isRequired,
  }).isRequired,
  allRoles: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      role: PropTypes.string,
    }),
  ).isRequired,
  editRole: PropTypes.func.isRequired,
  refetch: PropTypes.func.isRequired,
  currentPage: PropTypes.number.isRequired,
};

export default People;
