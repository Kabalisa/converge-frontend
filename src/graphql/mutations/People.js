import gql from 'graphql-tag';

const UPDATE_ROLES_MUTATION = gql`
  mutation changeUserRole($email: String!, $roleId: Int!) {
    changeUserRole(email: $email, roleId: $roleId){
      user{
        id
        email
        name
        roles{
          id
          role
        }
      }
    }
  }
`;

const INVITE_PERSON_MUTATION = gql`
mutation inviteToConverge($email: String!) {
  inviteToConverge(email: $email) {
    email
  }
}
`;

const CHANGE_USER_LOCATION_MUTATION = gql`
  mutation changeUserLocation($email: String!, $locationId: Int!) {
    changeUserLocation(email: $email, locationId: $locationId) {
      user {
        name
        location
      }
    }
  }
`;

const SET_USER_LOCATION = gql`
mutation setUserLocation($locationId: Int!) {
  setUserLocation(locationId: $locationId) {
    user {
      email
      location
    }
  }
}
`;

export {
  UPDATE_ROLES_MUTATION as default,
  INVITE_PERSON_MUTATION,
  CHANGE_USER_LOCATION_MUTATION,
  SET_USER_LOCATION,
};
