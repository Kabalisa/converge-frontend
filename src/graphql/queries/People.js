import gql from 'graphql-tag';

export const GET_PEOPLE_QUERY = gql`
query allUsers ($page: Int!, $perPage: Int!, $locationId: Int, $roleId: Int) {
  users(page: $page , perPage: $perPage, locationId: $locationId, roleId: $roleId){
    users{
      email
      name
      picture
      location
      id
      roles{
        role
        id
      }
    }
    pages
    hasNext
    hasPrevious
    queryTotal
  }
}
`;

const GET_USER_QUERY = gql`
query user ($email: String!) {
  user( email: $email) {
    id
    location
  }
}
`;

const GET_USER_ROLE = gql`
query user ($email: String!) {
  user( email: $email){
    roles{
      id
      role
    }
  }
 }
`;

const GET_ROLES_QUERY = gql`
query roles {
  roles {
    id
    role
  }
}`;

export { GET_PEOPLE_QUERY as default, GET_ROLES_QUERY, GET_USER_QUERY, GET_USER_ROLE };
