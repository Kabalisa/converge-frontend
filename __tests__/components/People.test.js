import React from 'react';
import { shallow } from 'enzyme';
import People from '../../src/components/People';

describe('People Component', () => {
  const person = {
    name: 'Edward Barton',
    location: 'Nairobi',
    accessLevel: 'Administrator',
  };
  const wrapper = shallow(<People people={person} />);
  it('renders properly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
