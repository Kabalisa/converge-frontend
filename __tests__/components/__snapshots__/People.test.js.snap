// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`People Component renders properly 1`] = `
<tr>
  <td>
    <img
      alt="profilePicture"
      className="profilePic"
      src="test-file-stub"
    />
    Edward Barton
  </td>
  <td>
    Nairobi
  </td>
  <td>
    Administrator
  </td>
  <td>
    <EditUser
      accessLevel="Administrator"
      userLocation="Nairobi"
      userName="Edward Barton"
    />
    <button>
      <img
        alt="Delete"
        src="test-file-stub"
      />
    </button>
  </td>
</tr>
`;
